File: backend/services/charging-service/src/main/java/com/chargingservice/services/ChargingService.java
// ===============================================================
// FILE: ChargingService.java (Phiên bản cập nhật)
// PACKAGE: com.chargingservice.services
// ===============================================================
package com.chargingservice.services;

import com.chargingservice.dtos.SessionResponseDto;
import com.chargingservice.dtos.StartSessionRequestDto;
import java.util.List;

public interface ChargingService {

    // --- CREATE ---
    SessionResponseDto startSession(StartSessionRequestDto requestDto);

    // --- UPDATE ---
    /**
     * Dừng một phiên sạc đang diễn ra.
     * @param sessionId ID của phiên sạc cần dừng.
     * @param requestDto Thông tin năng lượng đã sạc từ frontend
     * @return Thông tin chi tiết của phiên sạc đã hoàn thành.
     */
    SessionResponseDto stopSession(Long sessionId, com.chargingservice.dtos.StopSessionRequestDto requestDto);

    /**
     * Hủy một phiên sạc (ví dụ: người dùng hủy đặt chỗ hoặc hủy sạc).
     * @param sessionId ID của phiên sạc cần hủy.
     * @return Thông tin chi tiết của phiên sạc đã bị hủy.
     */
    SessionResponseDto cancelSession(Long sessionId);

    // --- READ ---
    /**
     * Lấy thông tin chi tiết của một phiên sạc.
     * @param sessionId ID của phiên sạc.
     * @return Thông tin chi tiết của phiên sạc.
     */
    SessionResponseDto getSessionById(Long sessionId);

    /**
     * Lấy danh sách tất cả các phiên sạc.
     * @return Danh sách các phiên sạc.
     */
    List<SessionResponseDto> getAllSessions();
    
    /**
     * Lấy trạng thái sạc real-time (SOC %, thời gian còn lại, chi phí)
     * @param sessionId ID của phiên sạc
     * @param speedMultiplier Tốc độ tua nhanh (1, 2, 4, 8) cho demo/testing
     * @return Trạng thái sạc chi tiết
     */
    com.chargingservice.dtos.SessionStatusDto getSessionStatus(Long sessionId, int speedMultiplier);
    
    /**
     * Đánh dấu session đã thanh toán
     * @param sessionId ID của phiên sạc
     * @param paymentId ID của payment record
     */
    void markSessionAsPaid(Long sessionId, Long paymentId);
}

File: frontend/src/services/chargingService.js
// src/services/chargingService.js
import apiClient from '../config/api';

/**
 * Lấy active session của user hiện tại
 * @returns {Promise<Object|null>} Session object nếu có, null nếu không có active session
 */
export const getActiveSession = async () => {
    try {
        const response = await apiClient.get('/sessions/active');
        // Nếu status 204 (no content), response.data có thể undefined
        if (response.status === 204 || !response.data) {
            return null;
        }
        return response.data;
    } catch (err) {
        // Nếu lỗi 204 (no content) từ axios interceptor
        if (err.response?.status === 204) {
            return null;
        }
        throw err;
    }
};

/**
 * Lấy thông tin session theo ID
 */
export const getSessionById = async (sessionId) => {
    const response = await apiClient.get(`/sessions/${sessionId}`);
    return response.data;
};

/**
 * Lấy trạng thái sạc real-time (SOC%, thời gian còn lại, chi phí)
 * @param {number} sessionId - ID của session
 * @param {number} speedMultiplier - Tốc độ tua nhanh (1, 2, 4, 8, 100) - default 1
 */
export const getSessionStatus = async (sessionId, speedMultiplier = 1) => {
    console.log(`🔍 API call: /sessions/${sessionId}/status?speedMultiplier=${speedMultiplier}`);
    const response = await apiClient.get(`/sessions/${sessionId}/status?speedMultiplier=${speedMultiplier}`);
    console.log('📥 Response SOC:', response.data?.currentSOC);
    return response.data;
};

/**
 * Kết thúc phiên sạc
 * @param {number} sessionId - ID của session
 * @param {object} stopData - {energyCharged, currentSOC} từ status hiện tại
 */
export const stopSession = async (sessionId, stopData = null) => {
    const response = await apiClient.post(`/sessions/${sessionId}/stop`, stopData);
    return response.data;
};

/**
 * Hủy phiên sạc
 */
export const cancelSession = async (sessionId) => {
    const response = await apiClient.post(`/sessions/${sessionId}/cancel`);
    return response.data;
};

